bindings/pyroot/JupyROOT/helpers/handlers.py

```diff
@@ -91,33 +113,39 @@ class Runner(object):
     Asynchronous
     >>> print(r.HasFinished())
     True
+    >>> r.Stop()
     '''
     def __init__(self, function):
         self.function = function
-        self.thread = None
+        self.is_running = False
+        self.argument_queue = queue.Queue()
+        self.poller = Poller(runner_obj=self, name = "JupyROOT Runner Thread")
+        self.poller.start()
+
+    def __del__(self):
+        if self.poller.is_alive():
+            self.Stop()
+        self.poller.join()
 
     def Run(self, argument):
         return self.function(argument)
 
     def AsyncRun(self, argument):
-        self.thread = Thread(target=self.Run, args =(argument,))
-        self.thread.start()
+        self.is_running = True
+        self.argument_queue.put(argument)
 
     def Wait(self):
-        if not self.thread: return
-        self.thread.join()
+        while self.is_running: pass
 
     def HasFinished(self):
-        if not self.thread: return True
-
-        finished = not self.thread.is_alive()
-        if not finished: return False
-
-        self.thread.join()
-        self.thread = None
-
+        if self.is_running: return False
         return True
 
+    def Stop(self):
+        self.Wait()
+        self.argument_queue.put(None)
+        self.Wait()
```

Comment:
Also you do not strictly need to wait here, right? In the end you will just join the poller thread when the `Runner` is destroyed.

Reply:
this is to be tried indeed.